File: go516/registers.go
package main

type Registers struct {
	read  [0x80]func() Word
	write [0x80]func(Word)
	value map[string]Word
	name  [0x80]struct {
		read  string
		write string
	}
	addr map[string]Addr
}

func NewRegisters() *Registers {
	return &Registers{
		value: make(map[string]Word),
		addr:  make(map[string]Addr),
	}
}

func (r *Registers) R(addr Addr, name string, fn func() Word) {
	r.read[addr] = fn
	r.name[addr].read = name
	r.addr[name] = addr
}

func (r *Registers) W(addr Addr, name string, fn func(Word)) {
	r.write[addr] = fn
	r.name[addr].write = name
	r.addr[name] = addr
}

func (r *Registers) Plain(addr Addr, name string) {
	r.value[name] = 0
	r.R(addr, name, func() Word {
		return r.value[name]
	})
	r.W(addr, name, func(val Word) {
		r.value[name] = val
	})
}

File: m516/web516/registers.js
if (typeof require != 'undefined') var hex = require('hex');

function Registers (size) {
	var dummy = {
		read : function () {return 0x0000},
		write : function (val) {},
	};
	
	for (var i=0;i<=size;i++) {
		this[i] = Object.create(dummy); // initialize registers to dummy
		this[i].name = {};
		this[i].name.read = 'unused';
		this[i].name.write = 'unused';
	}
	
	this.address = {};
}
Registers.prototype = [];

Registers.prototype.define = function (addr, type, name, fn) {
	this.address[name] = addr;
	this[addr][type] = fn;
	this[addr].name[type] = name;
}

Registers.prototype.R = function (addr, name, fn) {
	this.define(addr,'read',name,fn);
}

Registers.prototype.W = function (addr, name, fn) {
	let cb = (val) => {
		fn(val);
		if (this.onchange) {
			this.onchange(addr, name, val);
		}
	}
	this.define(addr,'write',name,cb);
}

Registers.prototype.plain = function (addr, name) {
	var r = this;
	this[name] = 0;
	this.R(addr,name,function(){return r[name]});
	this.W(addr,name,function(v){r[name] = v});
}

Registers.prototype.dump = function (from, to) {
	for (var i=from;i<=to;i++) {
		if (this[i].name.write != this[i].name.read) {
			console.log(hex(i,2) + ': ' + this[i].name.write + ' / ' + this[i].name.read);
		}
		else {
			console.log(hex(i,2) + ': ' + this[i].name.write);
		}
	}
}

if (typeof module != 'undefined' && module.exports) module.exports = Registers;